examples/vision/semantic_image_clustering.py

```diff
@@ -0,0 +1,576 @@
+        )
+        return loss
+
+
+"""
+### Implement representation learner
+"""
+
+
+def create_representation_learner(encoder, num_augumentations, projection_units):
+
+    inputs = keras.Input(shape=input_shape)
+    # Preprocess input images.
+    preprocessed = data_preprocessing(inputs)
+    # Create augmented versions of the images.
+    augmented = []
+    for _ in range(num_augumentations):
+        augmented.append(data_augmentation(preprocessed))
+    augmented = layers.Concatenate(axis=0)(augmented)
+    # Generate embedding representations of the images.
+    features = encoder(augmented)
+    # Add a projection head.
+    features = layers.Dropout(rate=0.1)(features)
+    x = layers.Dense(units=projection_units, use_bias=False)(features)
+    x = layers.BatchNormalization()(x)
+    outputs = layers.ReLU()(x)
+    # Create the model.
+    model = keras.Model(inputs=inputs, outputs=outputs, name="representation_learner")
+    return model
+
+
+"""
+### Train model
+"""
+
+encoder = create_encoder(representation_dim)
+representation_learner = create_representation_learner(
+    encoder, num_augumentations, projection_units
+)
+representation_learner.summary()
+
+# Since this self-supervised learning, we don't
+# use the y_data as our labels.
+labels = tf.ones(shape=(x_data.shape[0]))
+# Create a a Cosine decay learning rate scheduler.
+lr_scheduler = keras.experimental.CosineDecay(
+    initial_learning_rate=0.001, decay_steps=500, alpha=0.1
+)
+# Compile the model.
+representation_learner.compile(
+    optimizer=tfa.optimizers.AdamW(learning_rate=lr_scheduler, weight_decay=0.0001),
+    loss=ContrastiveLoss(temperature=0.1),
+)
+# Fit the model.
+history = representation_learner.fit(
+    x=x_data,
+    y=labels,
+    batch_size=512,
+    epochs=100,  # for better results, increase the number of epochs to 500.
+)
+
+"""
+Plot training loss
+"""
+
+plt.plot(history.history["loss"])
+plt.ylabel("loss")
+plt.xlabel("epoch")
+plt.show()
+
+"""
+## Compute nearest neighbours
+"""
+
+"""
+### Generate the embeddings for the images
+"""
+
+batch_size = 500
+# Get the feature vector representations of the images.
+feature_vectors = encoder.predict(x_data, batch_size=batch_size, verbose=1)
+# Normalize the feature vectores.
+feature_vectors = tf.math.l2_normalize(feature_vectors, -1)
+
+"""
+### Find the *k* nearest neighbours for each embedding
+"""
+
+neighbours = []
+num_batches = feature_vectors.shape[0] // batch_size
+for batch_idx in tqdm(range(num_batches)):
+    start_idx = batch_idx * batch_size
+    end_idx = start_idx + batch_size
+    current_batch = feature_vectors[start_idx:end_idx]
+    # Compute the dot similarity.
+    similarities = tf.linalg.matmul(current_batch, feature_vectors, transpose_b=True)
+    # Get the indices of most similar vectors.
+    _, indices = tf.math.top_k(similarities, k=k_neighbours + 1, sorted=True)
+    # Add the indices to the neighbours.
+    neighbours.append(indices[..., 1:])
```

Comment:
Here and elsewhere, I'd suggest to use US English spelling of the `neighbors` variable

Reply:
Agreed